### ISSUE TITLE
[training] 학습에 필요한 기본 클래스 구성

### ISSUE THREAD

**Issue by gowun** [sktaiflow/message-generator#3]
- data transformer: encoding, decoding with attention masks including dataloader
- bart message generator: forward, train, validate

**Comment by gowun** [comment]
https://github.com/sktaiflow/message-generator/blob/kennie/training/kennie/message_generator.py 

- argparse로 hyperparameter 조정하지 않고, 일반적인 function으로 조정하려는 시도 : *실패 ㅠ*
- 데이터 사이즈 5GB 이하일 경우, 멀티 GPU 사용의 의미가 없음
- 그러므로, 우리는 2개의 GPU로 2번의 다른 세팅의 학습을 하는 데 사용하는 것이 나음
- message_generator의 gpus는 사용 gpus 수가 아닌, 사용할 gpu 넘버 (integer)
- 만의 하나 multi-gpu를 쓰고 싶다면, gpus=-1(있는 거 다씀), auto_select_gpus=True(자동 세팅) 로 설정할 것
- custom loss function 붙이는 작업까지 끝나면 training/ 밑으로 옮길 예정

**Comment by gowun** [comment]
max_len에 따라 학습속도가 엄청나게 차이남?!!

**Comment by gowun** [comment]
특정한 조건(마지막 / 학습과 검증 간 loss 차이가 가장 적다던지 / 검증 로스가 가장 낮다던지)의 모델 저장 기능 추가
custom loss function 추가

**Comment by gowun** [comment]
https://github.com/sktaiflow/message-generator/blob/kennie/training/kennie/message_generator.py

function으로 학습 (모델 저장) 가능하게 변경하였으나, notebook에서는 구동하지 않아서 동 디렉토리의 do_train.py 처럼 구성해서 nohup python -u do_trian.py > ***.out & 으로 백그라운드로 돌리거나 터미널에서 파이썬 구동 후 돌리는 방법이 있음

**Comment by crystal0112** [comment]
케니님 generator기반으로 신규 생성된 토큰으로 학습 시키는 코드를 생성
확인 환료 되면 코드를 합칠 예정
https://github.com/sktaiflow/message-generator/blob/master/training/crystal/message_generator.py
